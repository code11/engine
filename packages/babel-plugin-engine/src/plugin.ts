import * as Babel from "@babel/core";
import { VariableDeclaratorVisitor } from "./visitors";

export type PluginConfig = {
  viewLibrary: string;
};

type pluginType = (babel: typeof Babel, state: PluginConfig) => {};

export const plugin: pluginType = (babel, state) => {
  return {
    name: "babel-plugin-engine",
    visitor: {
      VariableDeclarator: VariableDeclaratorVisitor(babel, state),
    },
  };
};
