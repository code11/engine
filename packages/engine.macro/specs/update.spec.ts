import pluginTester from "babel-plugin-tester";
import prettier from "prettier";
import plugin from "babel-plugin-macros";

const macroFile = "'@c11/engine.macro'";
const config = require("./babelConfig.json");

pluginTester({
  plugin,
  babelOptions: { filename: __filename, ...config },
  formatResult: (result: any) => {
    return prettier.format(result, {
      parser: "babel",
    });
  },
  tests: {
    "should support update": {
      code: `
        import { producer } from ${macroFile}
        const result: producer = ({
          a1 = update.foo,
          a2 = update[prop.foo],
          a3 = update[arg.a2],
          a4 = update.foo[prop.bar.baz],
          a5 = update.foo.bar[arg.a2],
          a6 = {
            baz: update.foo[arg.a4].baz,
          },
          a7 = update.foo[prop.a5.baz],
          a8 = update.foo[param.prop],
          a9 = update.foo[param.prop].baz,
          a10 = update.foo[param.prop].baz[prop.a2]
        }) => { }
      `,
      snapshot: true,
    },
  },
});
