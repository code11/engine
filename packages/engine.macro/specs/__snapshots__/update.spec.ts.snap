// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unknown plugin should support update: should support update 1`] = `

import { producer } from '@c11/engine.macro'
const result: producer = ({
  a1 = update.foo,
  a2 = update[prop.foo],
  a3 = update[arg.a2],
  a4 = update.foo[prop.bar.baz],
  a5 = update.foo.bar[arg.a2],
  a6 = {
    baz: update.foo[arg.a4].baz,
  },
  a7 = update.foo[prop.a5.baz],
  a8 = update.foo[param.prop],
  a9 = update.foo[param.prop].baz,
  a10 = update.foo[param.prop].baz[prop.a2]
}) => { }

      ↓ ↓ ↓ ↓ ↓ ↓

"use strict";

var result = {
  args: {
    type: "STRUCT",
    value: {
      a1: {
        type: "UPDATE",
        path: [
          {
            type: "CONST",
            value: "foo",
          },
        ],
      },
      a2: {
        type: "UPDATE",
        path: [
          {
            type: "EXTERNAL",
            path: ["foo"],
          },
        ],
      },
      a3: {
        type: "UPDATE",
        path: [
          {
            type: "INTERNAL",
            path: ["a2"],
          },
        ],
      },
      a4: {
        type: "UPDATE",
        path: [
          {
            type: "CONST",
            value: "foo",
          },
          {
            type: "EXTERNAL",
            path: ["bar", "baz"],
          },
        ],
      },
      a5: {
        type: "UPDATE",
        path: [
          {
            type: "CONST",
            value: "foo",
          },
          {
            type: "CONST",
            value: "bar",
          },
          {
            type: "INTERNAL",
            path: ["a2"],
          },
        ],
      },
      a6: {
        type: "STRUCT",
        value: {
          baz: {
            type: "UPDATE",
            path: [
              {
                type: "CONST",
                value: "foo",
              },
              {
                type: "INTERNAL",
                path: ["a4"],
              },
              {
                type: "CONST",
                value: "baz",
              },
            ],
          },
        },
      },
      a7: {
        type: "UPDATE",
        path: [
          {
            type: "CONST",
            value: "foo",
          },
          {
            type: "EXTERNAL",
            path: ["a5", "baz"],
          },
        ],
      },
      a8: {
        type: "UPDATE",
        path: [
          {
            type: "CONST",
            value: "foo",
          },
          {
            type: "INVOKE",
            path: ["prop"],
          },
        ],
      },
      a9: {
        type: "UPDATE",
        path: [
          {
            type: "CONST",
            value: "foo",
          },
          {
            type: "INVOKE",
            path: ["prop"],
          },
          {
            type: "CONST",
            value: "baz",
          },
        ],
      },
      a10: {
        type: "UPDATE",
        path: [
          {
            type: "CONST",
            value: "foo",
          },
          {
            type: "INVOKE",
            path: ["prop"],
          },
          {
            type: "CONST",
            value: "baz",
          },
          {
            type: "EXTERNAL",
            path: ["a2"],
          },
        ],
      },
    },
    meta: {},
  },
  fn: function fn(_ref) {
    var a1 = _ref.a1,
      a2 = _ref.a2,
      a3 = _ref.a3,
      a4 = _ref.a4,
      a5 = _ref.a5,
      a6 = _ref.a6,
      a7 = _ref.a7,
      a8 = _ref.a8,
      a9 = _ref.a9,
      a10 = _ref.a10;
  },
};


`;
